src/invite-api.js

```diff
@@ -1,236 +1,440 @@
+    }
+
+    this.#byProjectPublicId.delete(projectPublicId)
+
+    return pendingInvitesDeleted
   }
 }
 
+/**
+ * @typedef {Object} InviteApiEvents
+ * @property {(invite: Invite) => void} invite-received
+ * @property {(invite: Invite) => void} invite-removed
+ */
+
 /**
  * @extends {TypedEmitter<InviteApiEvents>}
  */
 export class InviteApi extends TypedEmitter {
-  // Maps project public id -> project id
-  /** @type {Map<string, string>} */
-  #projectIdsMapping = new Map()
-
-  // Maps project id -> set of device ids
-  /** @type {MapOfSets<string, string>} */
-  #peersToRespondTo = new MapOfSets()
-
-  // Maps project id -> { invite, fromPeerId }
-  /** @type {Map<string, { fromPeerId: string, invite: import('./generated/rpc.js').Invite }>} */
-  #pendingInvites = new Map()
-
   #isMember
   #addProject
+  #pendingInvites
 
   /**
    * @param {Object} options
    * @param {import('./local-peers.js').LocalPeers} options.rpc
    * @param {object} options.queries
-   * @param {(projectId: string) => Promise<boolean>} options.queries.isMember
-   * @param {(invite: import('./generated/rpc.js').Invite) => Promise<void>} options.queries.addProject
+   * @param {(projectId: string) => boolean} options.queries.isMember
+   * @param {(projectDetails: Pick<ProjectJoinDetails, 'projectKey' | 'encryptionKeys' | 'projectName'>) => Promise<unknown>} options.queries.addProject
    */
   constructor({ rpc, queries }) {
     super()
     this.rpc = rpc
     this.#isMember = queries.isMember
     this.#addProject = queries.addProject
 
-    this.rpc.on('invite', (peerId, invite) => {
-      this.#handleInvite(peerId, invite).catch(() => {
-        /* c8 ignore next */
-        // TODO: Log errors, but otherwise ignore them, but can't think of a reason there would be an error here
-      })
-    })
+    this.rpc.on('invite', this.#handleInvite)
+
+    this.#pendingInvites = new PendingInvites()
   }
 
   /**
    * @param {string} peerId
-   * @param {import('./generated/rpc.js').Invite} invite
+   * @param {InviteInternal} invite
    */
-  async #handleInvite(peerId, invite) {
-    const projectId = projectKeyToId(invite.projectKey)
-    const isAlreadyMember = await this.#isMember(projectId)
-
+  #handleInvite = (peerId, invite) => {
+    const isAlreadyMember = this.#isMember(invite.projectPublicId)
     if (isAlreadyMember) {
-      this.#sendAlreadyResponse({ peerId, projectId })
+      this.#sendAlreadyResponse({ peerId, inviteId: invite.inviteId })
       return
     }
 
-    const projectPublicId = projectKeyToPublicId(invite.projectKey)
-
-    this.#projectIdsMapping.set(projectPublicId, projectId)
-
-    const peerIds = this.#peersToRespondTo.get(projectId)
-    peerIds.add(peerId)
-
-    if (this.#pendingInvites.has(projectId)) return
+    const hasAlreadyReceivedThisInvite = this.#pendingInvites.hasInviteId(
+      invite.inviteId
+    )
+    if (hasAlreadyReceivedThisInvite) {
+      return
+    }
```

Comment:
Ah, I see you are doing the check here that you throw for in the pending invites state. I think we still need to catch errors in #handleInvite

Reply:
Addressed in d07b4a52ec3663f818aabdade188b1ed8bbe3a75.